mayastor/src/rebuild.rs

```diff
@@ -191,80 +224,115 @@ impl RebuildJob {
             state: RebuildState::Pending,
         })
     }
 
-    /// Rebuilds a non-healthy child from a healthy child from start to end
+    // Runs the management async task that kicks off N rebuild copy tasks and
+    // awaits each completion before kicking off more tasks until the
+    // bdev is fully rebuilt
     async fn run(&mut self) {
         self.change_state(RebuildState::Running);
-        self.current = self.start;
+        self.next = self.start;
         self.stats();
 
-        while self.current < self.end {
-            if let Err(e) = self.copy_one().await {
-                error!("Failed to copy segment {}", e);
-                self.change_state(RebuildState::Failed);
-                self.send_complete();
-                return;
-            }
-
-            if self.state == RebuildState::Stopped
-                || self.state == RebuildState::Paused
-            {
-                return self.send_complete();
+        self.start_all_tasks();
+        while self.tasks.active > 0 {
+            match self.await_one_task().await {
+                Some(r) => match r.error {
+                    None => {
+                        if self.state == RebuildState::Stopped
+                            || self.state == RebuildState::Paused
+                        {
+                            break;
+                        }
+
+                        self.start_task_by_id(r.id);
+                    }
+                    Some(e) => {
+                        error!("Failed to rebuild segment id {} block {} with error: {}", r.id, r.blk, e);
+                        self.change_state(RebuildState::Failed);
+                        // await all active tasks as we might still have ongoing
+                        // IO do we need a timeout?
+                        self.await_all_tasks().await;
+                        break;
+                    }
+                },
+                None => {
+                    // all senders have disconnected, out of place termination?
+                    self.change_state(RebuildState::Failed);
+
+                    if self.tasks.active != 0 {
+                        error!(
+                            "Completing rebuild with potentially {} active tasks",
+                            self.tasks.active
+                        );
+                    }
+                    break;
+                }
             }
         }
 
-        self.change_state(RebuildState::Completed);
-        self.send_complete();
+        self.complete();
     }
 
     /// Copies one segment worth of data from source into destination
-    async fn copy_one(&mut self) -> Result<(), RebuildError> {
-        // Adjust size of the last segment
-        if (self.current + self.segment_size_blks) >= self.start + self.end {
-            self.segment_size_blks = self.end - self.current;
-
-            self.copy_buffer = self
+    async fn copy_one(
+        &mut self,
+        id: u64,
+        current: u64,
```

Comment:
I wonder if "current" should be renamed to something like "start" or "start_blk". Within the scope of this function, current doesn't really mean anything; we are just copying a segment. This also matches the terminology used in the trace statement on line 297.

Reply:
Done